os/fs/smartfs/smartfs_smart.c

```diff
@@ -648,127 +650,224 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer, size_t
+				goto errout_with_semaphore;
 			}
 
-			/* Update our control variables */
-			sf->filepos += bytes;
-			sf->curroffset += bytes;
-			buflen -= bytes;
-			byteswritten += bytes;
+			header = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+			/* We work with a parallel branch but we use the tracking variables of the main branch
+			 * for correct referencing wrt to the total length of the file
+			 * This offset update may overwrite the offset update prior, and it is
+			 * correct since we are moving to a new sector
+			 */
+			sf->currsector = readwrite.logsector;
+			sf->curroffset = size;
 		}
+		byteswritten += bytes;
+		buflen -= bytes;
+		sf->filepos += bytes;
 
-		/* Test if we wrote to the end of the current sector */
-		if (sf->curroffset == fs->fs_llformat.availbytes) {
-			/* Wrote to the end of the sector. Update to point to the
-			 * next sector for additional writes. Read header to get the sector chain info.
-			 */
-			smartfs_setbuffer(&readwrite, sf->currsector, 0, size, (uint8_t *)fs->fs_rwbuffer);
-			ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+		fvdbg("Main Branch -> Main branch sector = %d, offset = %d, datalen = %d, filepos = %d\n", m_branch_sector, sf->curroffset, sf->entry.datalen, sf->filepos);
+		fvdbg("Parallel branch -> Start/Common sector = %u, Parallel branch sector = %d\n\n", overwrite_start_sector, p_branch_sector);
+
+		sf->bflags |= SMARTFS_BFLAG_DIRTY;
+	}
+
+	/* Check if we have more to overwrite
+	 * If yes, we will use the fs->rwbuffer while sf buffer holds on to the
+	 * updated conetns of the sector from where the overwrite branches out
+	 */
+	while (buflen > 0 && sf->filepos < sf->entry.datalen) {
+		fvdbg("OVERWRITING MORE, startng at PB sector %d, offset %d, main branch sector %d, buflen = %d\n\n", p_branch_sector, sf->curroffset, sf->currsector, buflen);
+		/* The number of bytes that wll be written is limited by
+		 * the minimum of the space in the current sector, the size of the file
+		 * or the buflen passed itself
+		 */
+		bytes = buflen;
+
+		if (bytes > (fs->fs_llformat.availbytes - sf->curroffset)) {
+			bytes = fs->fs_llformat.availbytes - sf->curroffset;
+		}
+
+		if (bytes > sf->entry.datalen - sf->filepos) {
+			bytes = sf->entry.datalen - sf->filepos;
+		}
+
+		/* Copy bytes from the input buffer to fs->rwbuffer */
+		memcpy(&fs->fs_rwbuffer[sf->curroffset], buffer, bytes);
+		smartfs_setbuffer(&readwrite, p_branch_sector, 0,\
+				fs->fs_llformat.availbytes, (uint8_t *)&fs->fs_rwbuffer[0]);
+
+		/* Will we be overwriting more bytes to a following sector?
+		 * If yes, pre-allocate a sector in the parallel branch and chain
+		 * it before writing the current sector.
+		 * However, before this, ssave the next chained sector in the main branch
+		 * for use later.
+		 */
+		if ((buflen - bytes) > 0 && (sf->filepos + bytes) < sf->entry.datalen) {
+			ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
```

Comment:
1. Do not apply this to master
2. Here all sector should be released.

Reply:
While we are allocating new sectors for creating a parallel branch with modified data,
We free the old chain after successfully committing the new chain in line 863

If we fail to commit the new chain because of a power outage, it is collected/released during recovery.

Reply:
No, it's totally wrong. I mean -ENOSPC, not power cut. Here all newly changed sector should be released. then they can be consumed for next operation, you should not wait until recovery.

Reply:
I understood your point now. Thankyou
I will add a fix here